### PR TITLE
Remove cookies, add publicKey.get, convert to factory function

### PR DESCRIPTION
**Pull request by Limess**
This library is currently difficult to use with `s3o-lambda` as simply requiring most of the useful methods start an `ft-poller` instance and immediately make a request - normally this is not what you want with Lambda functions.

This is a breaking change as the next release is likely to be a major due to ft-poller update requiring node.js >=8. 

#### What

Remove any methods specific to `express` rather than s3o. This can be handled in the relevant implementation; s3o middleware in this case.

Add cookie constants: name, token name and TTL, this means we have consistent cookies across node.js projects using s3o.

#### Why

There was a 2 year old TODO to action this on the file! The cookie setting/getting implementation for express is very different to Lambda, we benefit from having shared cookie names/TTLs in this library, but not the actual implementation.

#### What

Add a simple `get` method to retrieve public keys from s3o. This uses [got](https://github.com/sindresorhus/got) which handles retries and cache-control headers. If a `Map` instance is passed to this method an in-memory cache of the public key will be kept and the cache-control on the endpoint will be respected, if not a key will be fetched every time

#### Why

This is potentially useful for anyone implementing against s3o in node.js, rather than just `s3o-lambda` where this is extracted from.

#### What

Add a simple `get` method to retrieve public keys from s3o. This uses [got](https://github.com/sindresorhus/got) which handles retries and cache-control headers. If a `Map` instance is passed to this method an in-memory cache of the public key will be kept and the cache-control on the endpoint will be respected, if not a key will be fetched every time

#### Why

This is potentially useful for anyone implementing against s3o in node.js, rather than just `s3o-lambda` where this is extracted from.

#### What

Make `authentication.js` a factory function, a function which returns methods. This function now takes a function which returns an s3o publickey as an argument. This can be a poller or a regular public key getter.

Remove `s3oPublicKeyPromise`. This relied on s3o being eagerly instantiated on require as the poller needed to have been eagerly started. This can be immitated by either doing a public key fetch with `get` and a cache, or instantiating a poller.

#### Why

The eager initialisation of a poller fetching s3o tokens meant that this library wasn't useful outside of `s3o-middleware`. AWS lambda functions behaviour with polling is confusing as execution is frozen and resumed which may leave the poller mid-fetch, having it eagerly instantiated on require meant that lambdas would immediately start polling for new keys. This also made writing tests for express handlers using s3o-middleware awkward, as polling was started just by requiring the module under test.

Fixes https://github.com/Financial-Times/s3o-middleware-utils/issues/6

## Invisaged changes for `s3o-middleware`:

https://github.com/Financial-Times/s3o-middleware/blob/f3e59d380ce2c6a4f6fc47c01ef54a697ba2894c/index.js#L11-L12
```js
const {authenticate, publickey} = require('@financial-times/s3o-middleware-utils');
const { normaliseRequestCookies, setCookies, clearCookies } = require('../cookies');

const { authenticateToken, validate } = authenticate(publickey.poller(debug));

...


module.exports.validate = validate;
module.exports.ready = poller({promise: true}).then(function () { return true; });
```

Ideally `s3o-middleware` would also expose a factory function which then starts polling, but given the amount of consumers of that library I'm loath to make a breaking change to the API.

The code change itself won't be a breaking change in `s3o-middleware`, however this drops node 6 support from ft-poller (node 6 support ends very soon anyway) so would be a major bump.

### CODE REVIEW

**Review by JakeChampion — APPROVED** [review]

File: authenticate.js [hunk 1]
```diff
@@ -33,6 +41,10 @@ const authenticateToken = function (username, hostname, token) {
 	}
 };
 
-module.exports.authenticateToken = authenticateToken;
-module.exports.validate = validate;
-module.exports.s3oPublicKeyPromise = s3oPublicKey({ promise: true });
+module.exports = (getPublicKey) => {
+	const validate = validateFactory(getPublicKey);
+	return {
+		authenticateToken: authenticateToken(getPublicKey, validate),
```
Comment (wheresrhys) [inline authenticate.js]: name the factory function something slightly different? getTokenAuthenticator?

File: publickey.js [hunk 2]
```diff
@@ -11,7 +13,7 @@ const S3O_PUBLIC_KEY_URL = 'https://s3o.ft.com/publickey';
  * @return {function}       Returns a function that in turn returns either the public key as
  *                          as a string or a promise that resolves to the public key string.
  */
-module.exports = function (debug) {
+const poller = (debug) => {
 	let publicKey;
 
 	const flagsPoller = new Poller({
```
Comment (wheresrhys) [inline publickey.js:19]: not your error, but I suspect this was originaly copied from another library, 

rename to publicKeyPoller

File: test/authenticate.spec.js [hunk 3]
```diff
@@ -11,46 +11,50 @@ chai.should();
...
+	const authenticateFactory = proxyquire('../authenticate', {
 		'./validate': () => validateStub,
-	}).authenticateToken;
+	});
+	const getAuthenticateToken = () => authenticateFactory(publickeyPollerStub).authenticateToken;
 
 	beforeEach(() => {
-		publickeyStub.reset()
-		publickeyStub.returns(() => true);
+		publickeyPollerStub.reset();
```
Comment (wheresrhys) [inline test/authenticate.spec.js:22]: why is publickey used instead of publicKey everywhere in this file?

**Review by wheresrhys — APPROVED** [review]
a few comments. Too much context switching to dig into the detail, but what you propose in the description all sounds good

File: test/authenticate.spec.js (hunk 3 above)
Comment (Limess) [reply test/authenticate.spec.js:22]: Changed everywhere to `publickey` to match the export.

File: authenticate.js (hunk 1 above)
Comment (Limess) [reply authenticate.js]: Done!

File: publickey.js (hunk 2 above)
Comment (Limess) [reply publickey.js:19]: I've left this as is - it's in a `publicKey` file and exported as `poller` so it seems redundant.

**Pull request merged**